docs/dsa/binary_search/binary_search.md

```diff
@@ -0,0 +1,135 @@
+if (array[mid] == key) {
+return mid;
+} else if (array[mid] < key) {
+low = mid + 1;
+} else {
+high = mid - 1;
+}
+}
+return -1;
+}
+```
+
+How Binary Search Works
+Step-by-Step Explanation
+
+1.  Initialize: Set two pointers, low at the beginning and high at the end of the array.
+2.  Middle Element: Calculate the middle element's index.
+3.  Comparison:
+    If the middle element is the target, return its index.
+    If the middle element is less than the target, discard the left half by setting low to mid + 1.
+    If the middle element is greater than the target, discard the right half by setting high to mid - 1.
+4.  Repeat: Repeat steps 2 and 3 until the target is found or the low pointer exceeds the high pointer.
+
+Time Complexity
+The time complexity of binary search is
+𝑂
+(
+log
+⁡
+𝑛
+)
+O(logn), where
+𝑛
+n is the number of elements in the array. This logarithmic time complexity makes binary search significantly faster than linear search for large datasets.
+
+Practical Applications
+Binary search is widely used in various real-world applications and algorithmic problems:
+
+1. Searching in a Sorted Array
+   The primary use of binary search is to find elements in a sorted array efficiently. It is the foundation for more complex search algorithms.
+
+2. Dictionary Lookups
+   Binary search is used in dictionaries (like the one you're reading now) to quickly find words and their definitions.
+
+3. Binary Search Trees
+   Binary search is the basis for searching in binary search trees (BSTs), a fundamental data structure in computer science.
+
+4. Finding Boundaries
+   Binary search can be adapted to find the first or last occurrence of a target element, making it useful in problems requiring boundary searches.
+
+Variations of Binary Search
+Binary search can be adapted in various ways to solve different types of problems. Here are a couple of common variations:
+
+1. Lower Bound and Upper Bound
+   These variations of binary search are used to find the first and last occurrence of a target element in a sorted array.
+
+Lower Bound Pseudo-Code:
+
+```cpp
+FUNCTION lowerBound(array, key):
+low = 0
+high = array.length
+WHILE low < high:
+mid = (low + high) / 2
+IF array[mid] < key:
+low = mid + 1
+ELSE:
+high = mid
+RETURN low
+```
+
+Upper Bound Pseudo-Code:
+
+```cpp
+FUNCTION upperBound(array, key):
+low = 0
+high = array.length
+WHILE low < high:
+mid = (low + high) / 2
+IF array[mid] <= key:
+low = mid + 1
+ELSE:
+high = mid
+RETURN low
+```
+
+2. Rotated Sorted Array
+   Binary search can be modified to handle rotated sorted arrays, where the array is sorted but then rotated at some pivot point.
+
+Tips for Implementing Binary Search
+Handle Edge Cases: Ensure your implementation correctly handles cases where the target element is not present or when the array is empty.
+Prevent Overflow: When calculating the middle index, use mid = low + (high - low) / 2 instead of mid = (low + high) / 2 to prevent potential overflow.
```

Comment:
replace `mid = low + (high - low) / 2` to `$\text{mid} = \text{low} + \frac{\text{high} - \text{low}}{2}$`

Reply:
and also, `mid = (low + high) / 2` to `$\text{mid} = \frac{\text{low} + \text{high}}{2}$`